design/src/rfcs/rfc0037_retry_classifier_customization.md

```diff
@@ -0,0 +1,240 @@
+
+	println!("your buckets: {res:?}");
+
+    Ok(())
+}
+```
+
+### Customizing retry classification for an operation
+
+```rust
+#[tokio::main]
+async fn main() -> Result<(), aws_sdk_s3::Error> {
+    let sdk_config = aws_config::load_from_env().await;
+    let client = aws_sdk_s3::Client::new(&sdk_config);
+
+    let res = client
+	    .list_buckets()
+        .customize()
+        .await
+        .unwrap()
+        .config_override(   
+            aws_sdk_s3::Config::builder()
+	            .retry_classifier(CustomRetryClassifier)
+        )
+        .send()
+        .await?;
+
+	println!("your buckets: {res:?}");
+
+    Ok(())
+}
+```
+
+## How to actually implement this RFC
+
+In order to implement this feature, we must:
+- Update the current retry classification system so that individual classifiers as well as collections of classifiers can be easily composed together.
+- Create two new configuration mechanisms for users that allow them to customize retry classification at the service level and at the operation level.
+- Update retry classifiers so that they may 'short-circuit' the chain, ending retry classification immediately.
+
+### The `RetryClassifier` trait
+
+```rust
+/// The result of running a [`ClassifyRetry`] on a [`InterceptorContext`].
+#[non_exhaustive]
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum RetryClassifierResult {
+    /// "A retryable error was received. This is what kind of error it was,
+    /// in case that's important."
+    Error(ErrorKind),
+    /// "The server told us to wait this long before retrying the request."
+    Explicit(Duration),
+    /// "This response should not be retried."
+    DontRetry,
+}
+
+/// Classifies what kind of retry is needed for a given [`InterceptorContext`].  
+pub trait ClassifyRetry: Send + Sync + fmt::Debug {
+    /// Run this classifier on the [`InterceptorContext`] to determine if the previous request
+    /// should be retried. Returns a [`RetryClassifierResult`].  
+    fn classify_retry(
+        &self,
+        ctx: &InterceptorContext,
+        result_of_preceding_classifier: Option<RetryClassifierResult>,
+    ) -> Option<RetryClassifierResult>;
+
+    /// The name of this retry classifier.
+    ///
+    /// Used for debugging purposes
+    fn name(&self) -> &'static str;
+
+    /// The priority of this retry classifier. Classifiers with a higher priority will run before
+    /// classifiers with a lower priority. Classifiers with equal priorities make no guarantees
+    /// about which will run first.
+    fn priority(&self) -> RetryClassifierPriority {
+        RetryClassifierPriority::default()
+    }
+}
+```
+
+### Chaining retry classifiers
+
+Multiple retry classifiers are chained by wrapping classifiers inside one another. When classifiers must be wrapped in a specific order, use a specific type for the inner classifier. When classifiers must be composable in any order, use a referenced trait object. This approach is demonstrated in the following example code:
```

Comment:
Is this still accurate, or is it using priority-based sorting now?

Reply:
It's no longer accurate. I'll rework this section.